File: AdvancedSortingAlgorithm/CyclicSort/setMismatch.cpp
#include <iostream>
#include <vector>
using namespace std;


class Solution {
public:
    vector<int> findErrorNums(vector<int>& nums) {
        int n = nums.size();
        vector<int>ans(2,0);
        int i=0;
        while(i<n){
            int correctedIdx = nums[i]-1;
            if(nums[i]==i+1 || nums[i]==nums[correctedIdx])i++;
            else swap(nums[i],nums[correctedIdx]);
        }
        for(int i=0;i<n;i++){
            if(nums[i]!=i+1){
                ans[0]=nums[i];
                ans[1]=i+1;
                break;
            }
        }
        return ans;
    }
};

File: AdvancedSortingAlgorithm/CyclicSort/setMismatch.py
class Solution(object):
    def findErrorNums(self, nums):
        """
        :type nums: List[int]
        :rtype: List[int]
        """
        n = len(nums)
        ans = [0]*2
        i = 0
        while i < n:
            correctedIdx = nums[i]-1
            if nums[i] == i+1 or nums[correctedIdx] == nums[i]: 
                i+=1
            else:
                nums[i],nums[correctedIdx] = nums[correctedIdx],nums[i]
        for i in range(n):
            if nums[i] != i+1:
                ans[0] = nums[i]
                ans[1] = i+1
                break
        return ans
        